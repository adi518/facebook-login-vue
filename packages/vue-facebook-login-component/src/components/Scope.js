import {
  getSdk,
  login,
  logout,
  getLoginStatus,
  removeSdkScript
} from '@/helpers'

export default {
  name: 'v-facebook-login-scope',
  props: {
    value: {
      type: Object,
      default: () => ({ connected: false })
    },
    appId: {
      type: String,
      required: true
    },
    version: {
      type: String,
      default: 'v6.0' // https://developers.facebook.com/docs/graph-api/changelog/
    },
    options: {
      type: Object,
      default: () => ({})
    },
    loginOptions: {
      type: Object,
      default: () => ({ scope: 'email' })
    }
  },
  data: () => ({
    error: null,
    working: true,
    connected: false
  }),
  watch: {
    // eslint-disable-next-line no-unused-vars
    scope({ login, logout, toggleLogin, ...restScope }, _prevValue) {
      this.$emit('input', restScope)
    }
  },
  async beforeCreate() {
    await removeSdkScript()
  },
  async created() {
    const created = async () => {
      const sdk = await this.getSdk()
      if (this.error) return void 0
      this.$emit('sdk-init', { FB: sdk, scope: this.scope })
      const { status } = await getLoginStatus()
      if (status === 'connected') {
        this.connected = true
        this.$emit('login')
      }
    }
    this.async(created())
  },
  updated() {
    if (this.$slots.default?.length) {
      console.error('Slot must be scoped.')
    }
  },
  computed: {
    idle() {
      return !this.working
    },
    enabled() {
      return !this.disabled
    },
    disabled() {
      return this.working || Boolean(this.error) || !this.appId
    },
    disconnected() {
      return !this.connected
    },
    scope() {
      return {
        idle: this.idle,
        error: this.error,
        login: this.login,
        logout: this.logout,
        working: this.working,
        enabled: this.enabled,
        disabled: this.disabled,
        connected: this.connected,
        toggleLogin: this.toggleLogin,
        disconnected: this.disconnected
      }
    }
  },
  methods: {
    toggleLogin() {
      this.connected ? this.logout() : this.login()
    },
    async getSdk() {
      const { appId, version, options } = this
      const promise = getSdk({ appId, version, ...options })
      const sdk = await this.catchAsync(promise)
      return sdk
    },
    async login() {
      const promise = login(this.loginOptions)
      const { status } = await this.async(promise)
      if (status === 'connected') {
        this.connected = true
        this.$emit('login', status)
      }
      return promise
    },
    async logout() {
      await this.async(logout())
      this.connected = false
      this.$emit('logout')
      return logout
    },
    async catchAsync(promise) {
      return promise.catch(error => {
        if (error) {
          this.error = error
          console.error(this.error)
        }
      })
    },
    async async(promise) {
      this.working = true
      const handled = await this.catchAsync(promise)
      this.working = false
      return handled
    }
  },
  render() {
    if (this.$scopedSlots.default) {
      return this.$scopedSlots.default(this.scope)
    }
  }
}
