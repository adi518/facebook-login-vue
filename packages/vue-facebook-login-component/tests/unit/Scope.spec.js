import { shallowMount } from '@vue/test-utils'
import flushPromises from 'flush-promises'
import Scope from '@/components/Scope'

jest.mock('@/sdk', () => {
  const { Sdk } = jest.requireActual('@/sdk')
  return {
    __esModule: true,
    ...jest.requireActual('@/sdk'),
    Sdk: {
      isConnected: Sdk.isConnected,
      subscribe: jest.fn().mockResolvedValue({}),
      unsubscribe: jest.fn().mockResolvedValue()
    },
    getLoginStatus: jest.fn().mockResolvedValue({ status: 'unknown' }),
    login: jest.fn().mockResolvedValue({ status: 'connected' }),
    logout: jest.fn().mockResolvedValue()
  }
})

const { Sdk, getLoginStatus } = require('@/sdk')

const commonProps = { appId: '966242223397117' }

describe('Scope', () => {
  test('initial state and events (disconnected)', async () => {
    const wrapper = shallowMount(Scope, { propsData: commonProps })
    await flushPromises()
    const emitted = wrapper.emitted()
    expect(emitted['input']).toBeTruthy()
    expect(emitted['input'].length).toBe(1)
    expect(emitted['sdk-init']).toBeTruthy()
    expect(emitted['sdk-init'].length).toBe(1)
    expect(wrapper.emittedByOrder().map(e => e.name)).toEqual([
      'sdk-init',
      'input'
    ])
    expect(wrapper.vm.connected).toBe(false)
    expect(wrapper.isEmpty()).toBe(true)
  })

  test('initial state and events (connected)', async () => {
    getLoginStatus.mockResolvedValueOnce({ status: 'connected' })
    const wrapper = shallowMount(Scope, { propsData: commonProps })
    await flushPromises()
    const emitted = wrapper.emitted()
    expect(emitted['login']).toBeTruthy()
    expect(emitted['login'].length).toBe(1)
    expect(Object.keys(emitted).length).toBe(3)
    expect(wrapper.vm.connected).toBe(true)
    expect(wrapper.isEmpty()).toBe(true)
  })

  test('initial state with sdk error', async () => {
    Sdk.subscribe = jest.fn().mockRejectedValue('sdk error')
    Sdk.unsubscribe = jest.fn().mockImplementation()
    const wrapper = shallowMount(Scope, { propsData: commonProps })
    await flushPromises()
    expect(wrapper.vm.error).toBe('sdk error')
    expect(wrapper.isEmpty()).toBe(true)
  })

  test('login', async () => {
    const login = jest.fn()
    const wrapper = shallowMount(Scope, {
      propsData: commonProps,
      listeners: { login }
    })
    wrapper.vm.login()
    await flushPromises()
    expect(login).toHaveBeenCalledTimes(1)
    expect(login).toHaveBeenLastCalledWith({ status: 'connected' })
    expect(wrapper.isEmpty()).toBe(true)
  })

  test('logout', async () => {
    const logout = jest.fn()
    const wrapper = shallowMount(Scope, {
      propsData: commonProps,
      listeners: { logout }
    })
    wrapper.vm.logout()
    await flushPromises()
    expect(logout).toHaveBeenCalledTimes(1)
    expect(logout).toHaveBeenLastCalledWith()
    expect(wrapper.isEmpty()).toBe(true)
  })

  test('empty render with regular slot', async () => {
    const wrapper = shallowMount(Scope, {
      propsData: commonProps,
      slots: { default: '<div></div>' }
    })
    await flushPromises()
    expect(wrapper.isEmpty()).toBe(true)
  })

  test('render with scoped-slot', async () => {
    let scope
    const wrapper = shallowMount(Scope, {
      propsData: commonProps,
      scopedSlots: {
        default(props) {
          scope = props
          return <div>{props.connected ? 'connected' : 'disconnected'}</div>
        }
      }
    })
    await flushPromises()
    expect(wrapper.vm.scope).toBe(scope)
    expect(wrapper.element).toMatchSnapshot()
    wrapper.vm.connected = true
    wrapper.vm.$forceUpdate()
    await wrapper.vm.$nextTick()
    expect(wrapper.element).toMatchSnapshot()
  })
})
