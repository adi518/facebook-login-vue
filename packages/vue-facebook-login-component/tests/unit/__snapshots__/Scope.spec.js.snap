// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scope render with scoped-slot 1`] = `
<div>
  disconnected
</div>
`;

exports[`Scope render with scoped-slot 2`] = `
<div>
  connected
</div>
`;
